### PR TITLE
[REVIEW] Add support for column inputs in `strings::starts_with` and `strings::ends_with`

### PR DESCRIPTION
**Pull request by galipremsagar**
Resolves: #5074 

This PR adds support for `target` parameter to be `column_view` in libcudf. Additionally, cudf api `.str.startswith` & `.str.endswith` will also now support list-like inputs.

### CODE REVIEW

File: python/cudf/cudf/core/column/string.py [hunk 1]
```diff
@@ -2821,6 +2827,12 @@ def endswith(self, pat, **kwargs):
             result_col = column.column_empty(
                 len(self._column), dtype="bool", masked=True
             )
+        elif is_list_like(pat) or isinstance(
+            pat, (cudf.Series, cudf.Index, pd.Series, pd.Index)
+        ):
```
Comment (kkraus14) [inline python/cudf/cudf/core/column/string.py:2832]: May want to handle a `Column` here as well. Should probably put this into a utility function.

File: python/cudf/cudf/core/column/string.py [hunk 2]
```diff
@@ -2878,6 +2894,12 @@ def startswith(self, pat, **kwargs):
...
```
Comment (kkraus14) [inline python/cudf/cudf/core/column/string.py:2899]: Use utility function here as well

File: cpp/include/cudf/strings/find.hpp [hunk 3]
```diff
@@ -121,6 +121,28 @@ std::unique_ptr<column> starts_with(
   string_scalar const& target,
   rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
 
+/**
+ * @brief Returns a column of boolean values for each string where true indicates
+ * corresponding string in target column was found at the beginning of that string in
+ * the provided column.
+ *
+ * If `targets[i]` is not found at the beginning of a string in strings[i], false is set for
```
Comment (trevorsm7) [inline cpp/include/cudf/strings/find.hpp:129]:
```suggestion
 * If `targets[i]` is not found at the beginning of a string in `strings[i]`, false is set for
```

File: cpp/include/cudf/strings/find.hpp [hunk 4]
```diff
@@ -141,6 +163,27 @@ std::unique_ptr<column> ends_with(
   string_scalar const& target,
   rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
 
+/**
+ * @brief Returns a column of boolean values for each string where true indicates
+ * corresponding string in target column was found at the end of that string in
+ * the provided column.
+ *
+ * If `targets[i]` is not found at the end of a string in strings[i], false is set for
```
Comment (trevorsm7) [inline cpp/include/cudf/strings/find.hpp:171]:
```suggestion
 * If `targets[i]` is not found at the end of a string in `strings[i]`, false is set for
```

File: cpp/src/strings/find.cu [hunk 5]
```diff
@@ -216,6 +216,68 @@ std::unique_ptr<column> contains_fn(strings_column_view const& strings,
...
+ */
+template <typename BoolFunction>
+std::unique_ptr<column> contains_fn(strings_column_view const& strings,
+                                    strings_column_view const& targets,
+                                    BoolFunction pfn,
+                                    rmm::mr::device_memory_resource* mr,
+                                    cudaStream_t stream)
+{
+  auto strings_count = strings.size();
+  if (strings_count == 0) return make_numeric_column(data_type{BOOL8}, 0);
```
Comment (trevorsm7) [inline cpp/src/strings/find.cu:242]:
```suggestion
  if (strings_count == 0) return make_empty_column(data_type{BOOL8});
```

File: cpp/src/strings/find.cu [hunk 6]
```diff
@@ -216,6 +216,68 @@ std::unique_ptr<column> contains_fn(strings_column_view const& strings,
...
+    rmm::exec_policy(stream)->on(stream),
+    thrust::make_counting_iterator<size_type>(0),
+    thrust::make_counting_iterator<size_type>(strings_count),
+    d_results,
+    [d_strings, pfn, d_targets] __device__(size_type idx) {
+      // empty target string returns true
+      if (d_targets.is_valid(idx) && d_targets.element<string_view>(idx).length() == 0) {
+        return true;
+      } else if (!d_strings.is_null(idx) && !d_targets.is_null(idx)) {
+        return static_cast<bool>(
```
Comment (trevorsm7) [inline cpp/src/strings/find.cu:271]: Why `static_cast<bool>` the result of `pfn`? Alternatively `return bool{pfn(...)};` would emit a warning/error if `pfn` returns non-bool.

File: cpp/src/strings/find.cu [hunk 7]
```diff
@@ -286,5 +384,12 @@ std::unique_ptr<column> ends_with(strings_column_view const& strings,
   return detail::ends_with(strings, target, mr);
 }
 
+std::unique_ptr<column> ends_with(strings_column_view const& strings,
+                                  strings_column_view const& targets,
+                                  rmm::mr::device_memory_resource* mr)
+{
+  return detail::ends_with(strings, targets, mr);
```
Comment (trevorsm7) [inline cpp/src/strings/find.cu:391]:
```suggestion
  CUDF_FUNC_RANGE();
  return detail::ends_with(strings, targets, mr);
```

File: python/cudf/cudf/utils/dtypes.py [hunk 8]
```diff
@@ -203,9 +204,67 @@ def is_list_like(obj):
...
-    if isinstance(obj, (Sequence,)) and not isinstance(obj, (str, bytes)):
+    Returns
+    -------
+    Boolean: True or False depending on whether the
+    input `obj` is column-compatible or not.
+    """
+    if is_column_like(obj) or is_list_like(obj):
         return True
     else:
         return False
```
Comment (kkraus14) [inline python/cudf/cudf/utils/dtypes.py:270]:
```suggestion
    return is_column_like(obj) or is_list_like(obj)
```

File: python/cudf/cudf/utils/dtypes.py [hunk 9]
```diff
@@ -203,14 +204,69 @@ def is_list_like(obj):
...
+            and len(obj.__cuda_array_interface__["shape"]) == 1
+        )
+        or (
+            hasattr(obj, "__array_interface__")
+            and len(obj.__array_interface__["shape"]) == 1
+        )
+    ):
         return True
     else:
         return False
```
Comment (kkraus14) [inline python/cudf/cudf/utils/dtypes.py:250]: One more 😄
```suggestion
    return (
        isinstance(
            obj,
            (
                cudf.core.column.ColumnBase,
                cudf.Series,
                cudf.Index,
                pd.Series,
                pd.Index,
            ),
        )
        or (
            hasattr(obj, "__cuda_array_interface__")
            and len(obj.__cuda_array_interface__["shape"]) == 1
        )
        or (
            hasattr(obj, "__array_interface__")
            and len(obj.__array_interface__["shape"]) == 1
        )
    )
```

File: python/cudf/cudf/utils/dtypes.py [hunk 10]
```diff
@@ -203,14 +204,69 @@ def is_list_like(obj):
     Boolean: True or False depending on whether the
     input `obj` is like-like or not.
     """
-    from collections.abc import Sequence
+    if isinstance(obj, (Sequence,),) and not isinstance(obj, (str, bytes)):
+        return True
+    else:
+        return False
```
Comment (kkraus14) [inline python/cudf/cudf/utils/dtypes.py:210]:
```suggestion
    return isinstance(obj, (Sequence,),) and not isinstance(obj, (str, bytes))
```

**Pull request merged**